[ThreadContext] Update Async Execution Docs (#852)
SAP/cloud-sdk: docs/java/features/multi-tenancy/thread-context.mdx

Please note that the `RequestScopedHttpClientCache` which is the default being used by the SAP Cloud SDK cannot work with this approach.
:::

## Running Asynchronous Operations

As the name suggests, the `ThreadContext` is bound to a thread, more specifically to the one it was created in.
In consequence, other threads, i.e. asynchronous operations, cannot access the stored information if it hasn't been explicitly propagated.

The SAP Cloud SDK offers a dedicated API to execute asynchronous operations while preserving the `ThreadContext`.

Given any operation:

```java
Callable myTask = () -> doStuff();
// or
Runnable myTask = () -> doStuff();
```

The operation can be executed asynchronously via:

```java
Future runningTask = ThreadContextExecutors.submit(myTask);
```

> This will use a default `ThreadContextExecutorService` based on a `CachedThreadPool` to create a new Thread and run the task.

Reviewer: Next attempt for a clearer description 😅

```suggestion
Operations that are submitted this way will be executed by a single instance of `ThreadContextExecutorService`.
By default, this instance is based on a `CachedThreadPool`, which manages the concurrency internally.
```

:::tip Spring Integration
You can also use this functionality with Springs `@Async` annotation.
Refer to the [dedicated section below](#spring-integration).
:::

:::caution CAP Integration
In case you are using CAP, note that the CAP Request Context also needs to be passed on (in addition to the `ThreadContext`).
Please refer to [this guide](../../guides/cap-sdk-integration.mdx) on how to achieve this.
:::

### Spring Integration

You can conveniently integrate this functionality to work with Springs `@Async` annotation.

Below configuration will ensure all methods annotated with `@Async` will have the `ThreadContext` available:

```java
@EnableAsync
@Configuration
public class AsynchronousConfiguration implements AsyncConfigurer {

  @Override
  public Executor getAsyncExecutor() {
    return ThreadContextExecutors.getDefault();
  }
}
```

You can read more about the `@Async` functionality [here](https://www.baeldung.com/spring-async).

### Configuring the Executor

> The `ThreadContextExecutors` class leverages a default `ExecutorService` that can be configured.

Reviewer: ```suggestion
The `ThreadContextExecutors` class leverages a single `ThreadContextExecutorService` instance that can be configured.
```

```java
ThreadContextExecutorService executor = DefaultThreadContextExecutorService
            .of(Executors.newFixedThreadPool(3));

executor.submit(myTask);

// You can also set this executor as the default for ThreadContextExecutors:
ThreadContextExecutors.setDefault(executor);
ThreadContextExecutors.submit(myTask);
```

### Modifying the ThreadContext

You may want to run an asynchronous operation in a completely new or custom context.
By default, the executor will transfer any existing context to the new thread.

To run something in a completely new, empty context, use:

```java
ThreadContextExecutors.submit(myTask, new DefaultThreadContext());
```

By contrast, to pass on the current context, but modifying the tenant (and only the tenant), use:

```java
Tenant myTenant = new DefaultTenant("foo");
Callable myTaskWithTenant = () -> TenantAccessor.executeWithTenant(myTask, myTenant);

ThreadContextExecutors.submit(myTaskWithTenant);
```

To avoid multiple wrappings, in particular when changing multiple values, you can also build a custom executor:

```java
Tenant myTenant = new DefaultTenant("foo");
ThreadContextExecutor customExecutor = new ThreadContextExecutor()

>             .withListeners(new TenantThreadContextListener(myTenant));

Reviewer: Should we maybe modify a second value to make the example a little more complete?

// or
Runnable myTask = () -> doStuff();
```

The operation can be executed asynchronously via:

```java
Future runningTask = ThreadContextExecutors.submit(myTask);
```

Operations that are submitted this way will be executed by a single instance of `ThreadContextExecutorService`.
By default, this instance is based on a `CachedThreadPool`, which manages the concurrency internally.

:::tip Spring Integration
You can also use this functionality with Springs `@Async` annotation.
Refer to the [dedicated section below](#spring-integration).
:::

:::caution CAP Integration
In case you are using CAP, note that the CAP Request Context also needs to be passed on (in addition to the `ThreadContext`).
Please refer to [this guide](../../guides/cap-sdk-integration.mdx) on how to achieve this.
:::

### Spring Integration

You can conveniently integrate this functionality to work with Springs `@Async` annotation.

Below configuration will ensure all methods annotated with `@Async` will have the `ThreadContext` available:

```java
@EnableAsync
@Configuration
public class AsynchronousConfiguration implements AsyncConfigurer {

  @Override
  public Executor getAsyncExecutor() {

>     return ThreadContextExecutors.getDefault();

Reviewer 2: ```suggestion
    return ThreadContextExecutors.getExecutor();
```

   }
 }
 ```

You can read more about the `@Async` functionality [here](https://www.baeldung.com/spring-async).

### Configuring the Executor

The `ThreadContextExecutors` class leverages a default `ExecutorService` that can be configured.

You can create a custom `ThreadContextExecutorService`, for example to use a different thread pool, via:

```java
ThreadContextExecutorService executor = DefaultThreadContextExecutorService
            .of(Executors.newFixedThreadPool(3));

executor.submit(myTask);

// You can also set this executor as the default for ThreadContextExecutors:
ThreadContextExecutors.setDefault(executor);
ThreadContextExecutors.submit(myTask);
```

### Modifying the ThreadContext

You may want to run an asynchronous operation in a completely new or custom context.
By default, the executor will transfer any existing context to the new thread.

To run something in a completely new, empty context, use:
```java
ThreadContextExecutors.submit(myTask, new DefaultThreadContext());
```

By contrast, to pass on the current context, but modifying the tenant (and only the tenant), use:
```java
Tenant myTenant = new DefaultTenant("foo");

> ThreadContextExecutor customExecutor = new ThreadContextExecutor()

Author: TBD: Do we want to advertise this style? 

Alternatives would be:
```java
 Tenant myTenant = new DefaultTenant("foo");

Callable myTaskWithTenant = () -> TenantAccessor.executeWithTenant(myTask, myTenant);
ThreadContextExecutors.submit(myTaskWithTenant);
```
or:
```java
ThreadContext contextWithTenant = ThreadContextAccessor.getCurrentContext().duplicate().removeProperty("Tenant");
contextWithTenant.setPropertyIfAbsent("Tenant", Try.success(Property.of(myTenant)));

ThreadContextExecutors.submit(myTaskWithTenant, contextWithTenant);
```

Author: +1 on the third, I mentioned it mostly for completeness :D 

Additional downside of (1) is that the customer needs to interact with `ThreadContextExecutor` which is relatively easy to misuse (potentially less so if we apply [this improvement](https://github.wdf.sap.corp/MA/sdk/pull/7033#discussion_r4315070)).

We could also just describe both..

Reviewer: > We could also just describe both

Yeah, I think we should just mention both usages here.
Besides the technical "downsides" of the suggestion (2), I think it should be mentioned first since the semantic is just crystal clear and it is overall harder to get wrong.
Nevertheless, for more customizability (e.g. changing multiple values at once), suggestion (1) should be preferred.

Please note that the `RequestScopedHttpClientCache` which is the default being used by the SAP Cloud SDK cannot work with this approach.
:::

## Running Asynchronous Operations

> As the name suggests the `ThreadContext` is bound to a thread, more specifically to the one it was created.

Reviewer: ```suggestion
As the name suggests, the `ThreadContext` is bound to a thread, more specifically to the one it was created in.
```

> If asynchronous operations need to access the information, it has to be propagated to the new threads.

Reviewer: ```suggestion
In consequence, other threads, i.e. asynchronous operations, cannot access the stored information if it hasn't been explicitly propagated.
```

The SAP Cloud SDK offers a dedicated API to execute asynchronous operations while preserving the `ThreadContext`.

Given any operation:

```java
Callable myTask = () -> doStuff();
// or
Runnable myTask = () -> doStuff();
```

> The operation can be submitted via:

Reviewer: ```suggestion
The operation can be executed asynchronously via:
```

```java
Future runningTask = ThreadContextExecutors.submit(myTask);
```

> This will use a default `ThreadContextExecutorService` based on a `CachedThreadPool` to create a new Thread and run the task.

Reviewer: ```suggestion
By default, this will use a `ThreadContextExecutorService` based on a `CachedThreadPool` to create a new Thread and run the task.
```

Author: I would keep this as is,  to emphasise that this will always use "a default executor", namely the one set via `setDefault`

Reviewer: Okay, I see your point.

Nevertheless, I think the formulation _"a default executor"_ is not exactly describing what we want to convey here.
At least if I were reading this, I would ask myself _"What is a **default** executor? Is there a non-default one?"_.
Therefore, the actual method names `getDefault` and `setDefault` might not be optimal, as there is nothing else but this single "thingy", right?
At least I would assume if there is a `getDefault` that I somehow can also get a customized instance (similar to what the Apache HttpClient Builder is doing: You may use `HttpClients.default()` or `HttpClients.custom()`).

---

Overall, the most important part here (IMO) is that **all** submitted operations will be executed **in a single instance** of our interface.
And, yes, the instance that should be used to actually run the operations can be changed.
What do you think?

:::tip Spring Integration
You can also use this functionality with Springs `@Async` annotation.
Refer to the [dedicated section below](#spring-integration).
:::

:::caution CAP Integration
In case you are using CAP the CAP Request Context also needs to be passed on (in addition to the `ThreadContext`).
Please refer to [this guide](../../guides/cap-sdk-integration.mdx) on how to achieve this.
:::

### Spring Integration

You can conveniently integrate this functionality to work with Springs `@Async` annotation.

Below configuration will ensure all methods annotated with `@Async` will have the `ThreadContext` available:

```java
@EnableAsync
@Configuration
public class AsynchronousConfiguration implements AsyncConfigurer {

  @Override
  public Executor getAsyncExecutor() {
    return ThreadContextExecutors.getDefault();
  }
}

```

You can read more about the `@Async` functionality [here](https://www.baeldung.com/spring-async).

> ### Configuring the Executor

Author: I also thought about that, but found that people might not read on and miss the Spring integration, which is probably more relevant to most readers.

> The `ThreadContextExecutors` class leverages a default `ExecutorService` that can be configured.

Author: Yes, I only mention `ExecutorService` here to keep it as simple as possible. This is the thing that people are probably most interested in changing (and we do that as well in the example). I figured the detail is not too important and we can just leave the more specific subtype out here.

> The `ThreadContextExecutors` class leverages a default `ExecutorService` that can be configured.

Reviewer: Okay, fine with me.
